subnet/subnet_contract.go

```diff
@@ -210,6 +207,53 @@ func RemoveNodeAdmin(subnetID []byte, adminKey interop.PublicKey) {
 	storage.Delete(ctx, stNodeAdmKey)
 }
 
+// AddNode adds node to the specified subnetwork.
+// Must be called by subnet's owner or node administrator
+// only.
+func AddNode(subnetID []byte, node interop.PublicKey) {
+	if len(node) != interop.PublicKeyCompressedLen {
+		panic("addNode: " + ErrInvalidAdmin)
+	}
+
+	ctx := storage.GetContext()
+
+	stKey := append([]byte{ownerPrefix}, subnetID...)
+	prefixLen := len(stKey)
+
+	rawOwner := storage.Get(ctx, stKey)
+	if rawOwner == nil {
+		panic("addNode: " + ErrSubNotExist)
+	}
+
+	owner := rawOwner.([]byte)
+	if !runtime.CheckWitness(owner) {
+		var hasAccess bool
+
+		stKey[0] = nodeAdminPrefix
+
+		iter := storage.Find(ctx, stKey, storage.KeysOnly)
+		for iterator.Next(iter) {
+			key := iterator.Value(iter).([]byte)
+			if runtime.CheckWitness(key[prefixLen:]) {
+				hasAccess = true
+				break
+			}
+		}
+
+		if !hasAccess {
+			panic("addNode: " + ErrAccessDenied)
+		}
+	}
+
+	stKey[0] = nodePrefix
+
+	if keyInList(ctx, node, stKey) {
+		panic("addNode: node has already been added")
+	}
```

Comment:
yes, that is what `keyInList` does. it is quite a common operation, so I decided that it would be nice to have more explicit (mean naming) func for that

Reply:
yes, that is what `keyInList` does. it is quite a common operation, so I decided that it would be nice to have more explicit (mean naming) func for that

Reply:
yes, that is what `keyInList` does. it is quite a common operation, so I decided that it would be nice to have more explicit (mean naming) func for that